src/Microsoft.NetCore.Analyzers/Core/Security/DoNotUseCreateEncryptorWithNonDefaultIV.cs

```diff
@@ -0,0 +1,205 @@
+using Microsoft.CodeAnalysis.FlowAnalysis.DataFlow.PointsToAnalysis;
+using Microsoft.CodeAnalysis.Operations;
+using Microsoft.NetCore.Analyzers.Security.Helpers;
+
+namespace Microsoft.NetCore.Analyzers.Security
+{
+    [DiagnosticAnalyzer(LanguageNames.CSharp, LanguageNames.VisualBasic)]
+    public sealed class DoNotUseCreateEncryptorWithNonDefaultIV : DiagnosticAnalyzer
+    {
+        internal static DiagnosticDescriptor DefinitelyUseCreateEncryptorWithNonDefaultIVRule = SecurityHelpers.CreateDiagnosticDescriptor(
+            "CA5401",
+            nameof(MicrosoftNetCoreAnalyzersResources.DefinitelyUseCreateEncryptorWithNonDefaultIV),
+            nameof(MicrosoftNetCoreAnalyzersResources.DefinitelyUseCreateEncryptorWithNonDefaultIVMessage),
+            DiagnosticHelpers.EnabledByDefaultIfNotBuildingVSIX,
+            helpLinkUri: null,
+            descriptionResourceStringName: nameof(MicrosoftNetCoreAnalyzersResources.DoNotUseCreateEncryptorWithNonDefaultIVDescription),
+            customTags: WellKnownDiagnosticTagsExtensions.DataflowAndTelemetry);
+        internal static DiagnosticDescriptor MaybeUseCreateEncryptorWithNonDefaultIVRule = SecurityHelpers.CreateDiagnosticDescriptor(
+            "CA5402",
+            nameof(MicrosoftNetCoreAnalyzersResources.MaybeUseCreateEncryptorWithNonDefaultIV),
+            nameof(MicrosoftNetCoreAnalyzersResources.MaybeUseCreateEncryptorWithNonDefaultIVMessage),
+            DiagnosticHelpers.EnabledByDefaultIfNotBuildingVSIX,
+            helpLinkUri: null,
+            descriptionResourceStringName: nameof(MicrosoftNetCoreAnalyzersResources.DoNotUseCreateEncryptorWithNonDefaultIVDescription),
+            customTags: WellKnownDiagnosticTagsExtensions.DataflowAndTelemetry);
+
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(
+                                                                                        DefinitelyUseCreateEncryptorWithNonDefaultIVRule,
+                                                                                        MaybeUseCreateEncryptorWithNonDefaultIVRule);
+
+        private static readonly ConstructorMapper ConstructorMapper = new ConstructorMapper(
+            (IMethodSymbol constructorMethod, IReadOnlyList<PointsToAbstractValue> argumentPointsToAbstractValues) =>
+            {
+                return PropertySetAbstractValue.GetInstance(PropertySetAbstractValueKind.Unflagged);
+            });
+
+        private static readonly PropertyMapperCollection PropertyMappers = new PropertyMapperCollection(
+            new PropertyMapper(
+                "IV",
+                (PointsToAbstractValue pointsToAbstractValue) =>
+                {
+                    return PropertySetAbstractValueKind.Flagged;
+                }));
+
+        private static readonly HazardousUsageEvaluatorCollection HazardousUsageEvaluators = new HazardousUsageEvaluatorCollection(
```

Comment:
Did you decide not to take GenerateIV() into account?

Reply:
If not taking GenerateIV() into account, could probably skip PropertySetAnalysis and just produce a diagnostic on assigning to the .IV property.

---
In reply to: [323019864](https://github.com/dotnet/roslyn-analyzers/pull/2822#discussion_r323019864) [](ancestors = 323019864)

Reply:
I see what you mean now. But PropertyMapper can't tell if the property is set with a value from GenerateIV(). And if we just detect it's assignment, we will produce a diagnostic when there's no invocation of CreateEncryptor(). Is it ok?

Reply:
I'm sorry, I forgot about decryption!  We do need to care about if CreateEncryptor() is invoked.  Can you revert back to using PropertySetAnalysis?